File: securityinsecurebankv2/task3.py
#!python 2.7.18
#p3 failed. this code incorrect
import frida, sys

def on_message(message, data):
    print(message)

PACKAGE_NAME = "com.android.insecurebankv2"

jsscript = """
send("[FRIDA]-----Start Script-----");

Java.perform(function() {
	send("[FRIDA]-----Java.perform-----");
    var className = "com.android.insecurebankv2.ChangePassword";
    var methodName = "onCreate";
    var fieldName = "PASSWORD_PATTERN";

    var targetClass = Java.use(className);
    var targetMethod = targetClass[methodName];

    targetMethod.implementation = function(a) {
        console.log("........")
        try {
            // Change the value of the private static final string
            var fieldValue = "((?=.*\\d)(?=.*[a-z])(?=.*[@#$%]).{10,20})";
            var field = targetClass.class.getDeclaredField(fieldName);
            field.setAccessible(true);
            console.log("[Frida] Original value of " + fieldName + ": " + field.get(null));
            field.set(null, fieldValue);
            console.log("[Frida] Value of " + fieldName + " changed successfully!");
            console.log("[Frida] Modified value of " + fieldName + ": " + field.get(null));
        } catch (e) {
            console.error("[Frida] Error in changing " + fieldName + ":", e);
        }
        return this.onCreate(a);
    };
});
"""

process = frida.get_usb_device(1).attach(PACKAGE_NAME)
script = process.create_script(jsscript)
script.on('message', on_message)
print("[FRIDA]-----Running Hook-----")
script.load()
sys.stdin.read()

File: securityinsecurebankv2/task3.js
Java.perform(function() {
    var targetClassName = "com.android.insecurebankv2.ChangePassword";
    var targetFieldName = "PASSWORD_PATTERN";
    
    var targetClass = Java.use(targetClassName);
    
    // Hook the class initialization
    targetClass.$init.implementation = function() {
        // Call the original initialization
        this.$init.apply(this, arguments);

        try {
            // Use reflection to get the field
            var field = targetClass.class.getDeclaredField(targetFieldName);
            field.setAccessible(true);

            // Log the original value
            var originalValue = field.get(null);
            console.log("[Frida] Original value of " + targetFieldName + ": " + originalValue);

            // Modify the final modifier
            var modifiersField = java.lang.reflect.Field.class.getDeclaredField("modifiers");
            modifiersField.setAccessible(true);
            modifiersField.setInt(field, field.getModifiers() & ~java.lang.reflect.Modifier.FINAL);

            // Change the value of the private static final string
            field.set(null, "((?=.*\\d)(?=.*[a-z])(?=.*[A-Z])(?=.*[@#$%]).{10,20})");

            // Log the modified value
            var modifiedValue = field.get(null);
            console.log("[Frida] Modified value of " + targetFieldName + ": " + modifiedValue);

            console.log("[Frida] Value of " + targetFieldName + " changed successfully!");
        } catch (e) {
            console.error("[Frida] Error in changing " + targetFieldName + ":", e);
        }
    };
});